Build failure: python3Packages.imbalanced-learn

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

Build ``python3Packages.imbalanced-learn`` on master after https://github.com/NixOS/nixpkgs/pull/375144

Fails during runtime dependency check on hydra.

Once dependency is satisfied, fails during pytestCheckPhase.

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/290026763

### Relevant log output

```console
Sourcing python-remove-tests-dir-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing python-remove-bin-bytecode-hook.sh
Sourcing pypa-build-hook
Using pypaBuildPhase
Sourcing python-runtime-deps-check-hook
Using pythonRuntimeDepsCheckHook
Sourcing pypa-install-hook
Using pypaInstallPhase
Sourcing python-imports-check-hook.sh
Using pythonImportsCheckPhase
Sourcing python-namespaces-hook
Sourcing python-catch-conflicts-hook.sh
Sourcing pytest-check-hook
Using pytestCheckPhase
Running phase: unpackPhase
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking source archive /nix/store/5p1vls5pzlnhssbxxghh231qjrq15n9s-source
source root is source
setting SOURCE_DATE_EPOCH to timestamp 315619200 of file "source/references.bib"
Running phase: patchPhase
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: configurePhase
@nix { "action": "setPhase", "phase": "configurePhase" }
no configure script, doing nothing
Running phase: buildPhase
@nix { "action": "setPhase", "phase": "buildPhase" }
Executing pypaBuildPhase
Setting SETUPTOOLS_SCM_PRETEND_VERSION to 0.13.0
Including all tracked files automatically
Creating a wheel...
pypa build flags: --no-isolation --outdir dist/ --wheel
* Getting build dependencies for wheel...
running egg_info
creating imbalanced_learn.egg-info
writing imbalanced_learn.egg-info/PKG-INFO
writing dependency_links to imbalanced_learn.egg-info/dependency_links.txt
writing requirements to imbalanced_learn.egg-info/requires.txt
writing top-level names to imbalanced_learn.egg-info/top_level.txt
writing manifest file 'imbalanced_learn.egg-info/SOURCES.txt'
reading manifest file 'imbalanced_learn.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
adding license file 'LICENSE'
adding license file 'AUTHORS.rst'
writing manifest file 'imbalanced_learn.egg-info/SOURCES.txt'
* Building wheel...
running bdist_wheel
running build
running build_py
creating build/lib/imblearn
...skipping...
    
        # Test that the minimum dependencies in the README.rst file are
        # consistent with the minimum dependencies defined at the file:
        # pyproject.toml
    
        pyproject_path = Path(imblearn.__path__[0]).parents[0] / "pyproject.toml"
>       with open(pyproject_path, "rb") as f:
E       FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/75zslpzyq5srvv2xamp7l9680h3qgibd-python3.12-imbalanced-learn-0.13.0/lib/python3.12/site-packages/pyproject.toml'

imblearn/utils/tests/test_min_dependencies.py:28: FileNotFoundError
=============================== warnings summary ===============================
imblearn/utils/_test_common/instance_generator.py:117
imblearn/utils/_test_common/instance_generator.py:117
imblearn/utils/_test_common/instance_generator.py:117
imblearn/utils/_test_common/instance_generator.py:117
imblearn/utils/_test_common/instance_generator.py:117
  /nix/store/75zslpzyq5srvv2xamp7l9680h3qgibd-python3.12-imbalanced-learn-0.13.0/lib/python3.12/site-packages/imblearn/utils/_test_common/instance_generator.py:117: SkipTestWarning: Can't instantiate estimator >
    warnings.warn(msg, SkipTestWarning)

imblearn/combine/tests/test_smote_enn.py::test_sample_regular
  /nix/store/l5wkdd6rraj96y2nflfxqp33p9cv4wrg-python3.12-joblib-1.4.2/lib/python3.12/site-packages/joblib/externals/loky/backend/context.py:136: UserWarning: Could not find the number of physical cores for the >
  [Errno 2] No such file or directory: 'lscpu'
  Returning the number of logical cores instead. You can silence this warning by setting LOKY_MAX_CPU_COUNT to the number of cores you want to use.
    warnings.warn(

imblearn/ensemble/tests/test_bagging.py::test_oob_score_classification
imblearn/ensemble/tests/test_bagging.py::test_oob_score_classification
  /nix/store/qjgmb81jx90lj7c6xxrkq8i5h8rgjn7g-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/ensemble/_bagging.py:871: RuntimeWarning: invalid value encountered in divide
    oob_decision_function = predictions / predictions.sum(axis=1)[:, np.newaxis]

imblearn/ensemble/tests/test_bagging.py: 4 warnings
imblearn/ensemble/tests/test_easy_ensemble.py: 10 warnings
  /nix/store/l5wkdd6rraj96y2nflfxqp33p9cv4wrg-python3.12-joblib-1.4.2/lib/python3.12/site-packages/joblib/externals/loky/backend/fork_exec.py:38: DeprecationWarning: This process (pid=525) is multi-threaded, us>
    pid = os.fork()

imblearn/tests/test_common.py::test_estimators_compatibility_sklearn[FunctionSampler()-check_estimator_sparse_array]
imblearn/tests/test_common.py::test_estimators_compatibility_sklearn[FunctionSampler()-check_estimator_sparse_matrix]
  /nix/store/qjgmb81jx90lj7c6xxrkq8i5h8rgjn7g-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/utils/validation.py:1014: UserWarning: Can't check dok sparse matrix for nan or inf.
    array = _ensure_sparse_format(

imblearn/tests/test_common.py::test_estimators_compatibility_sklearn[Pipeline(steps=[('sampler',RandomUnderSampler(random_state=0)),('logistic',LogisticRegression())])-check_supervised_y_no_nan]
imblearn/tests/test_common.py::test_estimators_compatibility_sklearn[Pipeline(steps=[('sampler',RandomUnderSampler(random_state=0)),('logistic',LogisticRegression())])-check_supervised_y_no_nan]
imblearn/tests/test_common.py::test_estimators_compatibility_sklearn[RUSBoostClassifier()-check_supervised_y_no_nan]
imblearn/tests/test_common.py::test_estimators_compatibility_sklearn[RUSBoostClassifier()-check_supervised_y_no_nan]
  /nix/store/qjgmb81jx90lj7c6xxrkq8i5h8rgjn7g-python3.12-scikit-learn-1.6.1/lib/python3.12/site-packages/sklearn/utils/_array_api.py:399: RuntimeWarning: invalid value encountered in cast
    return x.astype(dtype, copy=copy, casting=casting)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=========================== short test summary info ============================
FAILED imblearn/tests/test_common.py::test_estimators_compatibility_sklearn[SMOTEN(random_state=42)-check_estimator_sparse_tag] - AssertionError: Estimator SMOTEN didn't fail when fitted on sparse data but...
FAILED imblearn/utils/tests/test_min_dependencies.py::test_min_dependencies_readme - FileNotFoundError: [Errno 2] No such file or directory: '/nix/store/75zslpz...
= 2 failed, 2527 passed, 37 skipped, 11 xfailed, 4 xpassed, 28 warnings in 108.00s (0:01:48) =
```

### Additional context
There is an abandoned commit for [sklearn-compat](https://github.com/NixOS/nixpkgs/pull/376175/commits/52270a87549818b860e02148225cf2a08a7dd622) which resolves the dep, but there is still 2 test failures.
[See](https://github.com/NixOS/nixpkgs/compare/master...philipwilk:nixpkgs:python3Packages.imbalanced-learn)

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.13.1, NixOS, 25.05 (Warbler), 25.05pre-git`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.25.5`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/gbvafd2vji5gy46210ri488asgn296hp-source`

### Notify maintainers

@rmcgibbo 
---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/development/python-modules/imbalanced-learn/default.nix
+++ b/pkgs/development/python-modules/imbalanced-learn/default.nix
@@ -14,6 +14,8 @@
   tensorflow,
   threadpoolctl,
   pytestCheckHook,
+  sklearn-compat,
+  python,
 }:
 
 buildPythonPackage rec {
@@ -41,6 +43,7 @@ buildPythonPackage rec {
     scikit-learn
     scipy
     threadpoolctl
+    sklearn-compat
   ];
 
   optional-dependencies = {
@@ -60,19 +63,35 @@ buildPythonPackage rec {
 
   preCheck = ''
     export HOME=$TMPDIR
+    # The GitHub source contains too many files picked up by pytest like
+    # examples and documentation files which can't pass.
+    cd $out/${python.sitePackages}
   '';
 
   disabledTestPaths = [
     # require tensorflow and keras, but we don't want to
     # add them to nativeCheckInputs just for this tests
-    "imblearn/keras/_generator.py"
+    "imblearn/keras"
+    "imblearn/tensorflow"
+    # even with precheck directory change, pytest still tries to test docstrings
+    "imblearn/tests/test_docstring_parameters.py"
+    # Skip dependencies test - pythonImportsCheck already does this
+    "imblearn/utils/tests/test_min_dependencies.py"
   ];
 
-  meta = with lib; {
+  disabledTests = [
+    # Broken upstream test https://github.com/scikit-learn-contrib/imbalanced-learn/issues/1131
+    "test_estimators_compatibility_sklearn"
+  ];
+
+  meta = {
     description = "Library offering a number of re-sampling techniques commonly used in datasets showing strong between-class imbalance";
     homepage = "https://github.com/scikit-learn-contrib/imbalanced-learn";
     changelog = "https://github.com/scikit-learn-contrib/imbalanced-learn/releases/tag/${version}";
-    license = licenses.mit;
-    maintainers = [ maintainers.rmcgibbo ];
+    license = lib.licenses.mit;
+    maintainers = with lib.maintainers; [
+      rmcgibbo
+      philipwilk
+    ];
   };
 }

--- a/pkgs/development/python-modules/sklearn-compat/default.nix
+++ b/pkgs/development/python-modules/sklearn-compat/default.nix
@@ -1,0 +1,52 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchFromGitHub,
+  hatchling,
+  scikit-learn,
+  pandas,
+  pytestCheckHook,
+  pytest-cov,
+  pytest-xdist,
+  pytz,
+}:
+
+buildPythonPackage rec {
+  pname = "sklearn-compat";
+  version = "0.1.3";
+  pyproject = true;
+
+  src = fetchFromGitHub {
+    owner = "sklearn-compat";
+    repo = "sklearn-compat";
+    tag = version;
+    hash = "sha256-9Py5VKDej7xDMdWVujkS/ooxOkEcjPJNlBHkbe/rycE=";
+  };
+
+  build-system = [
+    hatchling
+  ];
+
+  dependencies = [
+    scikit-learn
+  ];
+
+  nativeCheckInputs = [
+    pandas
+    pytestCheckHook
+    pytest-cov
+    pytest-xdist
+    pytz
+  ];
+
+  pythonImportsCheck = [
+    "sklearn_compat"
+  ];
+
+  meta = {
+    description = "Ease multi-version support for scikit-learn compatible library";
+    homepage = "https://github.com/sklearn-compat/sklearn-compat";
+    license = lib.licenses.bsd3;
+    maintainers = with lib.maintainers; [ philipwilk ];
+  };
+}

--- a/pkgs/top-level/python-packages.nix
+++ b/pkgs/top-level/python-packages.nix
@@ -16428,6 +16428,8 @@ self: super: with self; {
 
   skl2onnx = callPackage ../development/python-modules/skl2onnx { };
 
+  sklearn-compat = callPackage ../development/python-modules/sklearn-compat { };
+
   sklearn-deap = callPackage ../development/python-modules/sklearn-deap { };
 
   skodaconnect = callPackage ../development/python-modules/skodaconnect { };

